eclipse-che/che-operator
# Che Operator updates DWO and DWCO resources

Signed-off-by: Anatolii Bazko <abazko@redhat.com>



### What does this PR do?
- Che Operator sets annotations DWO and DWCO resources
- Che Operator syncs DWO and DWCO resources by updating them if hashes are different 
- Add DWO and DWCO tests for OCP

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

Che Operator does not update DWO and DWCO resources https://github.com/eclipse/che/issues/19664

### How to test this PR?

1. Run OCP cluster
2. Update ./local-debug.sh [1] by setting some old version of DWO and DWCO, for instance 
```
DEV_WORKSPACE_CONTROLLER_VERSION="v0.2.1"
DEV_WORKSPACE_CHE_OPERATOR_VERSION="v7.28.0"
```
3. Update CR [2] to turn off OAuth (spec.auth.openShiftoAuth: false)
4. Update CR [2] to turn on DevWorkspace (spec.devWorkspace.enable: true)
5. Run `./local-debug.sh`, wait until everything is deployed
6. Stop `./loca-debug.sh` and update DWO and DWCO, for instance:
```
DEV_WORKSPACE_CONTROLLER_VERSION="main"
DEV_WORKSPACE_CHE_OPERATOR_VERSION="main"
```
7. Run `./local-debug.sh`, wait until DWO and DWCO resources are updated. You can see that in the logs.
8. Check `devworkspace-che` and `devworkspace-controller` workspaces that all pods are running
9. Create workspaces and check that all pods are running.
```
oc apply -f https://raw.githubusercontent.com/che-incubator/devworkspace-che-operator/main/samples/flattened_theia-nodejs.yaml -n eclipse-che
oc apply -f https://raw.githubusercontent.com/devfile/devworkspace-operator/main/samples/flattened_theia-next.yaml -n eclipse-che
```

[1]  https://github.com/eclipse-che/che-operator/blob/main/local-debug.sh#L21-L22
[2] https://github.com/eclipse-che/che-operator/blob/main/deploy/crds/org_v1_che_cr.yaml

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
